File: Java/src/_0739_Daily_Temperatures/Solution.java
package _0739_Daily_Temperatures;

import java.util.ArrayDeque;
import java.util.Arrays;
import java.util.Deque;

/**
 * Date: 2020/9/25 14:40
 * Content:
 * 单调栈：单调递减的栈，一般单调栈存储的都是下标
 */
public class Solution {

    public int[] dailyTemperatures(int[] t) {

        int[] res = new int[t.length];
        Deque<Integer> stk = new ArrayDeque<>();
        for (int i = 0; i < t.length; i++) {
            while (!stk.isEmpty() && t[i] > t[stk.peek()]) {
                res[stk.peek()] = i - stk.peek();
                stk.pop();
            }
            stk.push(i);
        }

        return res;
    }

    public static void main(String[] args) {

        int[] t = {73, 74, 75, 71, 69, 72, 76, 73};
        System.out.println(Arrays.toString((new Solution()).dailyTemperatures(t)));
    }
}

File: Python/_0739_Daily_Temperatures/Solution.py
# coding=utf-8
# Date: 2021/7/17 16:23
from typing import List


# 执行用时：452 ms, 在所有 Python3 提交中击败了39.00%的用户
# 内存消耗：20 MB, 在所有 Python3 提交中击败了37.83%的用户
class Solution:
    def dailyTemperatures(self, t: List[int]) -> List[int]:
        n = len(t)
        res = [0] * n
        stk = []
        for i in range(n):
            while len(stk) != 0 and t[stk[len(stk) - 1]] < t[i]:
                res[stk[len(stk) - 1]] = i - stk[len(stk) - 1]
                stk.pop(len(stk) - 1)
            stk.append(i)
        return res


if __name__ == "__main__":
    print(Solution().dailyTemperatures([73, 74, 75, 71, 69, 72, 76, 73]))
